### PR TITLE
Fix JUnitEntryPoints annotation check

### LINKED ISSUE #1373: Found bug in JUnitEntryPoints
When using junit4 annotations, the static method `make(IClassHierarchy cha)` can not return any entrypoint.

I found that `isTestEntryPoint(TypeName typeName)` might be wrong:
```java
  private static boolean isTestEntryPoint(TypeName typeName) {
    // WALA uses $ to refers to inner classes. We have to replace "$" by "."
    // to make it a valid class name in Java source code.
    String javaName =
        StringStuff.jvmToReadableType(typeName.getPackage() + "." + typeName.getClassName());

    return TEST_ENTRY_POINT_ANNOTATION_NAMES.contains(javaName);
  }
```
For the class `Lorg/abc/Main`, `typeName.getPackage()` will return `org.abc` and delete the starting `L`. While `jvmToReadableType` depends on this `L` to parse the className. Without the starting `L`, `jvmToReadableType` will return an empty string.

So, I changed the method to the following:
```java
    private static boolean isTestEntryPoint(TypeName typeName) {
        // WALA uses $ to refers to inner classes. We have to replace "$" by "."
        // to make it a valid class name in Java source code.
        String javaName = StringStuff.jvmToReadableType(
                typeName.toString().replace('$', '.'));

        return TEST_ENTRY_POINT_ANNOTATION_NAMES.contains(javaName);
    }
```
And it works.

Note that in my case, I have no inner classes to test. So I just changed the code by the comment.

### PR DESCRIPTION
**Pull request by msridhar**
Fixes #1373

### CODE REVIEW

**Comment by msridhar** [comment]
@khatchad could you look this over?  You contributed the original JUnit entrypoint support.  Unfortunately we don't have tests and it's a bit of a pain to write one, so I'm not too sure about this change.

(@liblit: it's hard for me to write a test since I need allow the testSubjects code / source set in the core module to depend on junit, and I'm not too sure how to do that.  If you have any tips on that, that'd be welcome.)

**Comment by liblit** [comment]
> I need allow the testSubjects code / source set in the core module to depend on junit

I'm going to assume that you want JUnit 5, for which we already have [dependency versions defined in `gradle/libs.versions.toml`](https://github.com/wala/WALA/blob/09a71511f0d499a758d6ce08ffb394d5e3e04020/gradle/libs.versions.toml#L33-L38). It would also be possible to depend on JUnit 4, but that would require adding some lines to that same `gradle/libs.versions.toml` file.

Going with JUnit 5, I'm also going to guess that the dependency you want is the `junit-jupiter-api` artifact. I believe that's the JAR that contains the various annotations and other symbols used when authoring tests.

If my assumptions above are correct, then you can add the desired dependency by augmenting [the `dependencies` block in `core/build.gradle.kts`](https://github.com/wala/WALA/blob/09a71511f0d499a758d6ce08ffb394d5e3e04020/core/build.gradle.kts#L55-L68) as follows:

```kotlin
dependencies {
  ...
  "testSubjectsImplementation"(libs.junit.jupiter.api)
  "testSubjectsImplementation"(platform(libs.junit.bom))
}
```

**Important:** remember the double quotes around both uses of `testSubjectsImplementation`!

**Comment by msridhar** [comment]
Thanks a lot @liblit!  I'm now able to add a regression test and I have a lot more confidence in the fix here (in particular, the test fails before applying the fix).

File: core/src/test/java/com/ibm/wala/core/tests/cha/JunitEntrypointSupportTest.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import com.ibm.wala.util.collections.Iterator2Collection;
+import java.io.IOException;
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+public class JunitEntrypointSupportTest {
+
+  @Test
+  public void basic()
+      throws ClassHierarchyException, IllegalArgumentException, CancelException, IOException {
```
Comment (liblit) [inline core/src/test/java/com/ibm/wala/core/tests/cha/JunitEntrypointSupportTest.java:23]: I'm seeing a compilation failure reported on this line when using ECJ:

```
The declared exception CancelException is not actually thrown by the method basic() from type JunitEntrypointSupportTest
```

**Review by liblit — CHANGES_REQUESTED** [review]

File: core/src/test/java/com/ibm/wala/core/tests/cha/JunitEntrypointSupportTest.java (hunk 1 above)
Comment (msridhar) [reply core/src/test/java/com/ibm/wala/core/tests/cha/JunitEntrypointSupportTest.java]: Yup, my bad 🙂 Should be fixed now.

**Review by liblit — APPROVED** [review]

**Comment by khatchad** [comment]
> @khatchad could you look this over? You contributed the original JUnit entrypoint support. Unfortunately we don't have tests and it's a bit of a pain to write one, so I'm not too sure about this change.

I'm unsure where the 'L' is being removed at the reporter mentioned. [I am seeing it being added](https://github.com/wala/WALA/issues/1373#issuecomment-2135731893). But, if it is now being removed, these changes look fine to me.

File: core/src/main/java/com/ibm/wala/core/util/scope/JUnitEntryPoints.java [hunk 2]
```diff
@@ -132,8 +132,7 @@ public static Iterable<Entrypoint> make(IClassHierarchy cha) {
   private static boolean isTestEntryPoint(TypeName typeName) {
     // WALA uses $ to refers to inner classes. We have to replace "$" by "."
     // to make it a valid class name in Java source code.
-    String javaName =
-        StringStuff.jvmToReadableType(typeName.getPackage() + "." + typeName.getClassName());
+    String javaName = StringStuff.jvmToReadableType(typeName.toString().replace('$', '.'));
```
Comment (khatchad) [inline core/src/main/java/com/ibm/wala/core/util/scope/JUnitEntryPoints.java:135]: It would be good to have an inner class test as well.

**Review by khatchad — COMMENTED** [review]
A test method in an inner class would be a good test.

File: core/src/main/java/com/ibm/wala/core/util/scope/JUnitEntryPoints.java (hunk 2 above)
Comment (msridhar) [reply core/src/main/java/com/ibm/wala/core/util/scope/JUnitEntryPoints.java:135]: @khatchad can you sketch out a test case that would let us test this?  I'm not familiar with how nested classes are used with JUnit 4 / 5

Comment (khatchad) [reply core/src/main/java/com/ibm/wala/core/util/scope/JUnitEntryPoints.java:135]: I can have a look at this soon.

Comment (khatchad) [reply core/src/main/java/com/ibm/wala/core/util/scope/JUnitEntryPoints.java:135]: I think we just need to make sure we are getting the entry point corresponding to a method declared in a nested type.

Comment (khatchad) [reply core/src/main/java/com/ibm/wala/core/util/scope/JUnitEntryPoints.java:135]: https://github.com/wala/WALA/pull/1409

**Pull request merged**